app/main.py

```diff
@@ -39,51 +48,257 @@ def execute(self):
+                ) from exception
+        return {"result": response.json(), "status_code": response.status_code}
+
+    def convert_topology(self, event_type=0, event_timestamp=None):
+        """Function that will take care of update the shelve containing
+        the version control that will be updated every time a change is
+        detected in kytos topology, and return a new sdx topology"""
+        try:
+            # open the topology shelve
+            with shelve.open("topology_shelve") as open_shelve:
+                version = open_shelve['version']
+                self.dict_shelve = dict(open_shelve)  # pylint: disable=W0201
+                open_shelve.close()
+            if version >= 1 and event_type != 0:
+                timestamp = utils.get_timestamp()
+                if event_type == 1:
+                    version += 1
+                elif event_type == 2:
+                    timestamp = event_timestamp
+                return ParseConvertTopology(
+                    topology=self.get_kytos_topology(),
+                    version=version,
+                    timestamp=timestamp,
+                    model_version=self.dict_shelve['model_version'],
+                    oxp_name=self.dict_shelve['oxp_name'],
+                    oxp_url=self.dict_shelve['oxp_url'],
+                ).parse_convert_topology()
+            return {"result": topology_mock.topology_mock(),
+                    "status_code": 200}
+        except Exception as err:  # pylint: disable=W0703
+            log.info(err)
+            return {"result": "Validation Error", "status_code": 400}
+
+    def post_sdx_lc(self, event_type):
+        """ return the status from post sdx topology to sdx lc"""
+        post_topology = requests.post(
+                settings.SDX_LC_TOPOLOGY,
+                timeout=10,
+                json=self.sdx_topology)
+        if post_topology.status_code == 200:
+            if event_type != 0:
+                # open the topology shelve
+                with shelve.open("topology_shelve") as open_shelve:
+                    open_shelve['version'] = self.sdx_topology["version"]
+                    open_shelve['timestamp'] = self.sdx_topology["timestamp"]
+                    open_shelve['nodes'] = self.sdx_topology["nodes"]
+                    open_shelve['links'] = self.sdx_topology["links"]
+                    # now, we simply close the shelf file.
+                    open_shelve.close()
+            return {"result": self.sdx_topology,
+                    "status_code": post_topology.status_code}
+        return {"result": post_topology.json(),
+                "status_code": post_topology.status_code}
+
+    def post_sdx_topology(self, event_type=0, event_timestamp=None):
+        """ return the topology following the SDX data model"""
+        # pylint: disable=W0201
+        try:
+            if event_type != 0:
+                topology_update = self.convert_topology(
```

Comment:
Since you'll constantly need to have the full topology, another future optimization, is to maybe pass upfront the `topology` object, it's not that large, and then we could cut significantly the number of IO you're having to do, which might add extra load especially since you're subscribing to events that can be high frequency. For instance, a link flap on an interface, I bet would could result in hundreds of requests per second, I'll recommend that you also explore these cases to ensure it'll be prod grade. Anyway, if the client always need to the topology it might be worth augmenting topology events to also include the `topology` object too, other NApps that don't care can simply ignore it.

Reply:
Yes, we are currently assessing it with RENCI. Our aim is to efficiently transmit the topology to the SDX controller only when necessary and promptly execute actions. In cases where there is no need for a new topology, we will pass on the event and the related changes.

Reply:
Right. Since you're subscribing now to only `"kytos/topology.link_*", "kytos/topology.switch.*", kytos/topology.unloaded`" these aren't high frequency (link up has a timer), so the main concern that I raised has been indirectly addressed.